user/rcsput.man
.\" $Id: rcsput.man,v 11.9 2025/09/28 23:02:39 tom Exp $
.TH RCSPUT 1 2025-09-28 "" "User commands"
.ie n .ds CW R
.el   \{
.ie \n(.g .ds CW CR
.el       .ds CW CW
.\}
.de Es
.ne \\$1
.nr mE \\n(.f
.RS 5n
.sp .7
.nf
.nh
.ta 9n 17n 25n 33n 41n 49n
.ft \*(CW
..
.de Eh
.ft \\n(mE
.fi
.hy \\n(HY
.RE
.sp .7
..
.\" Bulleted paragraph
.de bP
.ie n  .IP \(bu 4
.el    .IP \(bu 2
..
.SH NAME
\fBrcsput\fR \-
RCS check-in utility
.SH SYNOPSIS
\fBrcsput\fI [options] [file-specifications]\fR
.SH DESCRIPTION
\fBRcsput\fR is a simple,
easy to use interface to \fIrcs\fR
(\fBr\fRevision \fBc\fRontrol \fBs\fRystem).
For each
file specified as input, it checks differences against the previously
archived version and prompts you for change history comments.
.PP
\fBRcsput\fR uses the \fIrcs\fR utility \fBci\fR to maintain
versions of a given source file in a dependent directory named "RCS":
.bP
It checks to ensure that each file is indeed a text
file (so that you do not accidentally archive ".o" files, for
example).
.bP
If you give \fBrcsput\fR a directory name, it
will recur, checking-in files in the directory.
.bP
For each file which has a corresponding ",v" file,
\fBrcsput\fR compares the two (using \fBdiff\fR) and pipes the result
through the pager.
.bP
An option is provided so that you may direct \fBrcsput\fR
to perform the differencing without checking the file into rcs.
.bP
The ",v" file is post-processed by \fBrcsput\fR
so that the check-in date matches the file's modification date.
.PP
The last point is the fundamental advantage offered by \fBrcsput\fR.
The ordinary \fIrcs\fR methodology uses the current date as the
check-in date.
This works well only for large projects in which a
central project administrator is responsible for controlling the versions
of source files.
It does not work well for small projects,
for which \fIrcs\fR's primary advantage
is its compact storage of multiple versions of a file.
.PP
By using the file's modification date as a reference, you can more
easily back up to a meaningful version \- by date, rather than
version number.
.PP
\fBRcsput\fR integrates all of the functions used in the \fIrcs\fR
check-in process into one utility program.
.SH OPTIONS
Some of the options which you may specify to \fBrcsput\fR are
passed through to the underlying \fBci\fR utility.
Others represent extensions:
.TP 4n
.B \-b
is passed to \fBdiff\fR, and directs it
to ignore trailing blanks on a line, and to treat repeated blanks
as a single blank.
.TP 5
.B \-c
directs \fBrcsput\fR to use \fBcat\fR rather
than the \fBPAGER\fR (usually \fBmore\fR) to display differences.
This is most useful in an Apollo pad, since the \fBmore\fR program
would otherwise switch to VT100 emulator mode.
.TP 5
.B \-d
instructs \fBrcsput\fR to test for differences,
but not to check the files into \fIrcs\fR.
.TP 5
.B \-h
is passed to \fBdiff\fR, and permits it to handle
huge differences.
.TP 5
.BI \-L file
causes \fBrcsput\fR to generate a
log-file of the files which are processed, and all differences which
are encountered.
The log-file is inherited in recursion to lower
directory levels (i.e., it is written to the same place).
If no argument
is specified, rcsput assumes "logfile".
.TP
.BI \-T path
specifies an alternate tool to invoke,
overrides the default "\fBcheckin\fR".
.SH OPERATIONS
The \fBrcsput\fR utility is designed for use in small development
projects.
The methodology for this tool follows:
.bP
Develop source files "normally".
Each file should
contain rcs keywords (see \fIci\ (1)\fR) so that you will be
able to distinguish checked-out files.
The rcs keywords should appear at the top of your source file, for consistency.
In C language programs,
the convention is to make a string which will permit the \fBident\fR
utility to show the versions of the modules which make up a program:
.Es 3
#ifndef	lint
static char ident[] = "$\&Id: rcsput.man,v 10.1 92/02/06 10:01:45 dickey Exp $";
#endif
.Eh
.bP
Periodically archive (with \fBrcsput\fR) those versions
of files which you wish to keep
(you should never have programs which have new features
which you wish to keep, while there are defects in other parts of
the program.
That would be an unsound approach to development!).
.bP
When you reach the point of releasing the program, ensure
that all source files have been checked-in.
The directory editor
(\fBded\fR) is useful for reviewing the check-in dates.
.bP
Copy the directory containing your program to the release
directory.
Purge all files, except those which are stored in the \fIrcs\fR subdirectories.
Use \fBrcsget\fR to extract the files.
The unadorned \fBco\fR utility will work, of course,
but it retains the file modification dates.
You may also use \fBcheckout\fR to retain file dates.
.bP
Ensure that all files have been checked-in and released.
You may use \fBdiff\fR to compare the directories \- the only
differences should be the substituted \fIrcs\fR keywords.
.bP
Build the released version of your program.
All files should be present.
No embedded path names should refer to your development copy.
To ensure good isolation, you may change the permissions on
your development directory temporarily.
.PP
When checking files into \fIrcs\fR, it is a good idea to make
a test run (using the "\-\fBd\fR" option) so that you can inspect
the differences.
For example, you may have forgotten to remove (or bypass) debugging stubs.
Or, you may have been editing a checked-out
file (with the \fIrcs\fR keywords substituted).
\fBRcsput\fR would archive this anyway.
.SH ENVIRONMENT
\fBRcsput\fR is written in C, and runs on POSIX systems.
.PP
Environment variables imported by \fBrcsput\fR include:
.TP
.B PAGER
identifies the program to use in displaying differences
between the file which is being checked in, and the previously archived
version.
There may be a lot of differences \- more than can be
shown on one screen.
.SH FILES
\fBRcsput\fR uses the following files
.TP
.B checkin
A utility which invokes \fBci\fR, and modifies
the \fIrcs\fR ",v" file after check-in so that the check-in
date matches the file's modification date.
.SH ANTICIPATED CHANGES
None.
.SH AUTHORS
Thomas E. Dickey <dickey@invisible-island.net>
.SH SEE ALSO
checkin, rcsget, checkout, ded, ci\ (1), co\ (1), diff\ (1),
ident\ (1)


CHANGES
2025-09-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* package/debian/copyright: formatting

	* user/baseline.man, user/checkin.man, user/checkout.man,
	  user/checkup.man, user/copy.man, user/link2rcs.man, user/permit.man,
	  user/rcsput.man:
	check-manpage

	* user/baseline.man, user/checkin.man, user/checkout.man,
	  user/checkup.man, user/copy.man, user/link2rcs.man, user/permit.man,
	  user/rcsput.man:
	fixup-th-man

	* user/rcsget.man, user/vcs.man: check-manpage

	* user/rcsget.man, user/vcs.man: fixup-th-man

	* configure: regen

	* aclocal.m4, configure.in: use CF_PROG_INSTALL

2025-07-09  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: update to 2025-07-10

2025-06-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: fix license

2025-01-17  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING, package/cm_tools-12.0.spec,
	  package/debian/changelog:
	bump

	* configure: regen

	* configure.in: use CF_WITH_SYSTYPE early

2025-01-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/checkin/src/checkin.c, src/checkout/src/checkout.c,
	  src/checkup/src/checkup.c, src/checkout/src/rcsget.c,
	  src/copy/src/copy.c, src/link2rcs/src/link2rcs.c,
	  src/permit/src/permit.c, src/vcs/src/delete.c, src/vcs/src/unlock.c,
	  src/checkin/src/rcsput.c, src/baseline/src/baseline.c:
	gcc15-warnings

2024-12-21  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

2024-07-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2024/07/27

2024-06-03  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2024/05/27

2024-04-25  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING, package/cm_tools-12.0.spec,
	  package/debian/changelog:
	bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2023-12-27  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2023-12-27

2023-12-02  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2023-12-02

2023-01-22  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec: bump

	* configure: regen

	* bin/makefile.in:
	build-fix for Solaris: rename variable $(USER) to $(AS_USER)

	* package/debian/copyright, COPYING: bump

	* aclocal.m4, configure.in: use pkg-config to find td_lib

	* package/debian/changelog: bump

2023-01-20  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/compat: bump

2022-12-31  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* aclocal.m4: build-fixes for NetBSD

	* configure: regen

2022-10-11  Thomas E. Dickey  <dickey@invisible-island.net>

	* user/makefile.in: use install-sh if needed

	* package/cm_tools-12.0.spec, package/debian/changelog,
	  package/debian/copyright, COPYING:
	bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2022-08-01  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2022-08-01

2021-03-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2021-01-10  Thomas E. Dickey  <dickey@invisible-island.net>

	* configure: regen

	* aclocal.m4: quoting fixes for shellcheck

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* COPYING, package/cm_tools-12.0.spec, package/debian/changelog,
	  package/debian/copyright:
	bump

2020-12-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* user/makefile.in, src/vcs/src/makefile.in,
	  src/baseline/test/makefile.in, src/baseline/makefile.in,
	  src/checkin/test/makefile.in, src/checkout/src/makefile.in,
	  src/checkout/test/makefile.in, src/checkup/src/makefile.in,
	  src/checkup/test/makefile.in, src/copy/src/makefile.in,
	  src/copy/test/makefile.in, src/link2rcs/src/makefile.in,
	  src/link2rcs/test/makefile.in, support/makefile.in:
	use VPATH

2020-12-07  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/checkin/makefile.in, src/checkout/makefile.in,
	  src/checkup/makefile.in, src/copy/makefile.in,
	  src/link2rcs/makefile.in, src/permit/makefile.in, src/vcs/makefile.in:
	use VPATH

	* package/debian/compat: bump

	* src/baseline/src/makefile.in, src/checkin/src/makefile.in,
	  src/permit/src/makefile.in, src/permit/test/makefile.in,
	  src/vcs/test/makefile.in:
	use VPATH

2020-12-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* configure.in: without-X

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

2020-11-30  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: update to 2020-11-14

2020-08-16  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2020-08-17

2019-12-21  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* user/checkup.man: fix typo

2019-12-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* user/rcsput.man, user/rcsget.man, user/permit.man, user/link2rcs.man,
	  user/checkup.man, user/checkin.man, user/copy.man, user/checkout.man,
	  user/baseline.man:
	check-manpage fixes (rename Ex to Eh for consistency, other errata)

2019-12-05  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* src/checkin/src/rcsput.c, src/checkin/src/checkin.c:
	use DYN-argv lists to replace catarg()

	* src/vcs/src/vcs.c, src/permit/src/permit.c:
	use show_argv2(), to handle verb vs argv[0]

	* src/checkout/src/checkout.c, src/checkout/src/rcsget.c:
	use DYN-argv lists to replace catarg()


src/checkin/src/rcsput.c
/*
 * Title:	rcsput.c (rcs put-tree)
 * Author:	T.E.Dickey
 * Created:	19 Oct 1989
 * Modified:
 *		05 Dec 2019, use DYN-argv lists.
 *		13 Jan 2013, pass "-M" option to checkin.
 *		10 May 1997, pass-thru -w option to checkin.
 *		09 May 1994, port to Linux
 *		22 Sep 1993, gcc warnings
 *		04 Mar 1992, test "-f" after 'different()'
 *		05 Feb 1992, revised filename-parsing with 'rcsargpair()',
 *			     obsoleted "-x".
 *		04 Feb 1992, pass piped-in text via "-m" option.
 *		11 Oct 1991, converted to ANSI
 *		01 Oct 1991, added "-B" option for 'checkin'
 *		13 Sep 1991, moved 'filesize()' to common-lib
 *		20 Jun 1991, use 'shoarg()'
 *		06 Jun 1991, use "-x" option in local name-checking
 *		03 Jun 1991, pass-thru "-x" to 'checkin'
 *		20 May 1991, mods to compile on apollo sr10.3
 *		19 Apr 1990, added "-T" option (to permit non-checkin tool use)
 *		18 Apr 1990, modified call on rcs2name/name2rcs to support "-x"
 *			     option in checkin/checkout
 *		06 Dec 1989, added interpretation of "-f" (force) option to
 *			     override test-for-diffs, test-for-textfile, and
 *			     test-for-existence of arguments.  Also, interpret
 *			     "-?" argument to show not only usage message from
 *			     this utility, but from checkin too.
 *		01 Nov 1989, walktree passes null pointer to stat-block if
 *			     no-access.
 *
 * Function:	Use 'checkin' to checkin one or more files from the
 *		RCS-directory which is located in the current working
 *		directory, and then, to set the delta date of the  checked-in
 *		files according to the last modification date (rather than the
 *		current date, as RCS assumes).
 *
 * Options:	see 'usage()'
 */

#define	CUR_PTYPES
#define	STR_PTYPES
#include	<ptypes.h>
#include	<rcsdefs.h>
#include	<sccsdefs.h>
#include	<dyn_str.h>
extern char *tmpnam(char *);

MODULE_ID("$Id: rcsput.c,v 11.14 2025/01/07 00:49:12 tom Exp $")

#define	VERBOSE		if (!quiet) PRINTF

static ARGV *ci_opts;
static ARGV *diff_opts;
static const char *verb = "checkin";
static FILE *log_fp;
static int a_opt;		/* all-directory scan */
static int no_op;		/* no-op mode */
static const char *pager;	/* nonzero if we don't cat diffs */
static int force;
static int quiet;

static void
cat2fp(FILE *fp, char *name)
{
    FILE *ifp;
    char t[BUFSIZ];
    size_t n;

    if ((ifp = fopen(name, "r")) != NULL) {
	while ((n = fread(t, sizeof(char), sizeof(t), ifp)) > 0)
	    if (fwrite(t, sizeof(char), n, fp) != n)
		  break;
	FCLOSE(ifp);
    }
}

static int
different(const char *working)
{
    ARGV *cmds;
    ARGV *opts;
    DYN *flat;
    static const char *prog = "rcsdiff";

    FILE *ifp, *ofp;
    char buffer[BUFSIZ], out_diff[BUFSIZ];
    int changed = FALSE;
    size_t n;

    opts = argv_init();
    argv_merge(&opts, diff_opts);
    argv_append(&opts, working);

    if (!quiet)
	show_argv2(stdout, prog, argv_values(opts));

    /* kludgey, but we don't do many pipes */
    cmds = argv_init1(prog);
    argv_merge(&cmds, opts);

    if (!tmpnam(out_diff))
	failed("tmpnam");
    if (!(ofp = fopen(out_diff, "w")))
	failed("open-tmpnam");

    flat = argv_flatten(cmds);
    if (!(ifp = popen(dyn_string(flat), "r")))
	failed("popen");
    dyn_free(flat);

    /* copy the result to a file so we can send it two places */
    while ((n = fread(buffer, sizeof(char), sizeof(buffer), ifp)) > 0) {
	if (fwrite(buffer, sizeof(char), n, ofp) != n)
	      break;
	changed = TRUE;
    }
    (void) pclose(ifp);
    FCLOSE(ofp);

    if (changed) {
	if (!quiet) {
	    if (pager == NULL)
		cat2fp(stdout, out_diff);
	    else {
		if (execute(pager, out_diff) < 0)
		    failed(pager);
	    }
	}
	if (log_fp != NULL) {
	    PRINTF("appending to logfile");
	    cat2fp(log_fp, out_diff);
	}
    } else if (!quiet) {
	PRINTF("*** no differences found ***\n");
    }

    (void) unlink(out_diff);
    argv_free(&opts);
    argv_free(&cmds);
    return (changed);
}

static void
checkin(const char *path, const char *working, const char *archive)
{
    ARGV *args;
    int first;

    if ((first = (filesize(archive) < 0)) != 0) {
	if (!force && !istextfile(working)) {
	    PRINTF("*** \"%s\" does not seem to be a text file\n",
		   working);
	    return;
	}
	first = TRUE;
    } else {
	time_t date;
	const char *vers, *locker;
	rcslast(path, working, &vers, &date, &locker);
	if (*vers == '?')
	    first = TRUE;	/* no revisions present */
	else {
	    if (!different(working) && !force)
		return;
	    first = FALSE;
	}
    }

    args = argv_init1(verb);
    argv_merge(&args, ci_opts);
    argv_append(&args, working);
    argv_append(&args, archive);

    if (!no_op) {
	PRINTF("*** %s \"%s\"\n",
	       first ? "Initial RCS insertion of"
	       : "Applying RCS delta to",
	       working);
    } else {
	PRINTF("--- %s \"%s\"\n",
	       first ? "This would be initial for"
	       : "Delta would be applied to",
	       working);
    }

    if (!quiet)
	show_argv2(stdout, verb, argv_values(args));
    if (!no_op) {
	if (executev(argv_values(args)) < 0)
	    failed(working);
    }
    argv_free(&args);
}

/*
 * Test for directories that we don't try to scan
 */
static int
ignore_dir(char *path)
{
    if (sameleaf(path, sccs_dir((char *) 0, path))
	|| sameleaf(path, rcs_dir((char *) 0, path))) {
	if (!quiet)
	    PRINTF("...skip %s\n", path);
	return TRUE;
    }
    return FALSE;
}

/*ARGSUSED*/
static int
WALK_FUNC(scan_tree)
{
    char tmp[BUFSIZ], *s = pathcat(tmp, path, name);

    (void) level;

    if (RCS_DEBUG)
	PRINTF("++ scan %s / %s\n", path, name);

    if (sp == NULL || readable < 0) {
	readable = -1;
	if (!ignore_dir(s)) {	/* could be RCS-dir we cannot scan */
	    perror(name);
	    if (!force)
		exit(FAIL);
	}
    } else if (isDIR(sp->st_mode)) {
	abspath(s);		/* get rid of "." and ".." names */
	if (!a_opt && *pathleaf(s) == '.')
	    readable = -1;
	else if (ignore_dir(s))
	    readable = -1;
	else
	    track_wd(path);
    } else if (isFILE(sp->st_mode)) {
	track_wd(path);
	checkin(path, name, name2rcs(name, FALSE));
    } else
	readable = -1;

    return (readable);
}

static void
do_arg(const char *name)
{
    (void) walktree((char *) 0, name, scan_tree, "r", 0);
}

static void
usage(int option)
{
    static const char *tbl[] =
    {
	"Usage: rcsput [options] files_or_directories"
	,""
	,"Options include all CHECKIN-options, plus:"
	,"  -a       process all directories, including those beginning with \".\""
	,"  -b       (passed to rcsdiff)"
	,"  -h       (passed to rcsdiff)"
	,"  -c       send differences to terminal without $PAGER filtering"
	,"  -d       compute differences only, don't try to CHECKIN"
	,"  -L file  write all differences to logfile"
	,"  -T TOOL  specify alternate tool to \"checkin\" to invoke per-file"
	,""
    };
    unsigned j;
    for (j = 0; j < sizeof(tbl) / sizeof(tbl[0]); j++)
	FPRINTF(stderr, "%s\n", tbl[j]);
    if (option == '?')
	(void) system("checkin -?");
    exit(FAIL);
}

/*ARGSUSED*/
_MAIN
{
    int j;
    const char *s;
    char *cat_input = NULL;
    int m_opt = FALSE;
    char original[MAXPATHLEN];

    if (!isatty(fileno(stdin)) && interactive())
	cat_input = strtrim(file2mem("-"));

    track_wd((char *) 0);
    pager = dftenv("more -l", "PAGER");

    if (!getwd(original))
	failed("getwd");

    ci_opts = argv_init();
    diff_opts = argv_init();

    /* process options */
    for (j = 1; (j < argc) && (*(s = argv[j]) == '-'); j++) {
	if (strchr("BqrfklumMnNstw", s[1]) != NULL) {
	    argv_append(&ci_opts, s);
	    switch (s[1]) {
	    case 'f':
		force = TRUE;
		break;
	    case 'm':
		m_opt = TRUE;
		break;
	    case 'q':
		quiet = TRUE;
		argv_append(&diff_opts, s);
		break;
	    }
	} else {
	    switch (s[1]) {
	    case 'a':
		a_opt = TRUE;
		break;
	    case 'b':
	    case 'h':
		argv_append(&diff_opts, s);
		break;
	    case 'c':
		pager = NULL;
		break;
	    case 'd':
		no_op = TRUE;
		break;
	    case 'L':
		if (s[2] == EOS)
		    s = "logfile";
		if (!(log_fp = fopen(s, "a+")))
		    usage(0);
		break;
	    case 'T':
		verb = s + 2;
		break;
	    default:
		usage(s[1]);
	    }
	}
    }

    if (cat_input && !m_opt)
	argv_append2(&ci_opts, "-m", cat_input);

    /* process list of filenames */
    if (j < argc) {
	while (j < argc) {
	    char working[MAXPATHLEN];
	    char archive[MAXPATHLEN];
	    Stat_t sb;

	    j = rcsargpair(j, argc, argv);
	    if (rcs_working(working, &sb) < 0)
		failed(working);

	    if (isDIR(sb.st_mode)) {
		if (!ignore_dir(working))
		    do_arg(working);
	    } else {
		(void) rcs_archive(archive, (Stat_t *) 0);
		checkin(original, working, archive);
	    }
	}
    } else
	do_arg(".");

    exit(SUCCESS);
    /*NOTREACHED */
}
